[#29] - Posts CRUD endpoints

Dj5822 opened the pull request:
# Description:
Updated the database such that the posts table has the following:
text_content, author, parent, usersLiked, usersShared, createdAt, updatedAt, attachments (not currently set up properly)

Created posts.js which contains the following API endpoints:
GET posts
POST posts
PUT posts
DELETE posts

Also set up the plan for the unit tests, however they have not been implemented yet.

Note that certain functionality does not work yet, for example when the parent post is deleted, children post are not also deleted yet.

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](../CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] My code follows the code style of this project
- [x] My code has been commented
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

Dj5822 committed:
Modified the posts table in the database.

Dj5822 committed:
GET post is working

Dj5822 committed:
CRUD methods for posts setup (no auth)

Dj5822 committed:
Merge branch 'main' into posts-setup

Dj5822 committed:
Authentication working for posts

Dj5822 committed:
Testing plan

Dj5822 committed:
Added comments to test file

LemonPepperSeasoning commented on backend/server/controllers/posts.js:
```diff
@@ -0,0 +1,152 @@
+            */
+            const createNewPost = await models.posts.create({
+                text_content: body.text_content,
+                author: 'test author',
```
I think this line should be "body.author" not "test author" :satisfied:

Dj5822 replied:
I changed it to author: decodedUser.username
I'm not sure if the author is suppose to be the author's id or the author's username though.

LemonPepperSeasoning reviewed (commented):
Very minor change is needed for "backend/server/controllers/posts.js".

The author is hard coded to be "test author".

Dj5822 committed:
Updated createPost

Fixed errors:
1) author is correctly posted.
2) multiple response messages were being sent, fixed it to send only one response message.

mdar365 commented on backend/database/migrations/create-posts.js:
```diff
@@ -1,19 +1,38 @@
+module.exports = {
+    up: (queryInterface, Sequelize) =>
+        queryInterface.createTable('posts', {
+            id: {
+                allowNull: false,
+                autoIncrement: true,
+                primaryKey: true,
+                type: Sequelize.INTEGER,
+            },
+            text_content: {
+                type: Sequelize.STRING,
+            },
+            author: {
+                type: Sequelize.STRING,
+            },
+            parent: {
+                type: Sequelize.INTEGER,
+            },
+            usersLiked: {
+                type: Sequelize.INTEGER,
+            },
+            usersShared: {
+                type: Sequelize.INTEGER,
+            },
+            createdAt: {
+                allowNull: false,
+                type: Sequelize.DATE,
+            },
+            updatedAt: {
+                allowNull: false,
+                type: Sequelize.DATE,
+            },
+            attachments: {
+                type: Sequelize.STRING,
+            },
+        }),
```
The actual model and the response we send to FE is slightly different, the real model, the one we save to the DB should match this:
**Post**
* id
* content (string)
* author (userID)
* parent (ID of parent post)
* children (array of post IDs)
* usersLiked (array of IDs of users who likes the post)
* usersShared (array of IDs users who shared the post)
* timestamp (unix timestamp indicating timestamp of creation)
* Attachments (link to attachment in Firebase db)

After you do all the processing in your controllers, you should convert the object you're about to send as a response to this format:
![image](https://user-images.githubusercontent.com/68842181/157351472-84962b45-0c8e-45e2-a91f-64de0aae5f7e.png)

mdar365 replied:
You might need to create a postsDTO class and have a method that creates a postsDTO object from the whole post object

Dj5822 replied:
Right now attachments is a seperate issue, so I don't think I need to fix that yet. 
As for usersLiked and usersShared, I'm going to leave those as integers for testing purposes (at least for now).

Dj5822 replied:
I fixed the author field though.

mdar365 replied:
Agreed, posts models can be finalised in a separate issue, for now we just need the CRUD endpoints

mdar365 commented on backend/server/controllers/posts.js:
```diff
@@ -0,0 +1,150 @@
+*/
+export const getPostById = async (req, res) => {
+    try {
+        const { params } = req
```
Should check the request is authorized first (all endpoints require authorization except the create users endpoint)

Dj5822 replied:
I think it will be better to not have authentication for viewing posts

mdar365 replied:
Maybe discuss it with FE first then, they said they wanted everything to have authentication

Dj5822 replied:
Michael said "not for now probs", so I'm just going to resolve this conversation.

mdar365 replied:
yeah allg

mdar365 commented on backend/server/controllers/posts.js:
```diff
@@ -0,0 +1,150 @@
+        } else {
+            const { params, body } = req
+
+            const updated = await models.posts.update(
```
Need to check post being updated belongs to the decodedUser, same with deletePost

Dj5822 replied:
[Checks whether the author can edit or delete post](https://github.com/SE701Team2/Updog/pull/38/commits/e288bf3cba48adaec01af18b9778a72a0576bca8)
[e288bf3](https://github.com/SE701Team2/Updog/pull/38/commits/e288bf3cba48adaec01af18b9778a72a0576bca8)

mdar365 commented on backend/server/routes/index.js:
```diff
@@ -1,14 +1,32 @@
+*/
+router.route('/posts').post(posts.createPost)
+
+router.route('/posts/:id').get(posts.getPostById)
+
+router.route('/posts/:id').put(posts.modifyPostById)
+
+router.route('/posts/:id').delete(posts.deletePostById)
```
You could do 
`router.route('/posts/:id').get(posts.getPostById).put(posts.modifyPostById).delete(posts.deletePostById)` since they all have the same url

Dj5822 replied:
Fixed this in [Cleaned up routes](https://github.com/SE701Team2/Updog/pull/38/commits/f8480556fada553aca484cff47a40350fdbc0a15)
[f848055](https://github.com/SE701Team2/Updog/pull/38/commits/f8480556fada553aca484cff47a40350fdbc0a15).

Dj5822 committed:
Added tests for GET /posts

Dj5822 committed:
Added tests for POST /posts

Dj5822 committed:
Merge branch 'posts-setup' of https://github.com/Dj5822/Updog into posts-setup

Dj5822 committed:
Changed author field to author id

Dj5822 committed:
Cleaned up routes

Dj5822 committed:
Checks whether the author can edit or delete post

Dj5822 committed:
Checks whether the parent of a post is valid

Dj5822 committed:
Added tests

mdar365 reviewed (approved):
Looks good, but I noticed you keep sending requests to the endpoints for creating a user or generating authentication token in your tests when you could've just called the functions directly. such as using models.users.create() or maybe even new models.users to create a new user object. I recommend changing them if you have time

mdar365 commented:
Actually yeah it's important to switch to using the actual functions instead, because if you call the endpoints then any changes to the endpoints could ruin your tests. Only endpoints you should be calling are the ones you're testing

Dj5822 committed:
Removed usage of /api/users endpoints from tests
